### ISSUE TITLE
initial version of property automation

### ISSUE THREAD

**Issue by ahicks92** [synthizer/synthizer#73]
One common use case of Synthizer is to be able to set up fades and things like that.  Let's add the ability to do simple automation on any property.  Initially, we'll support block-level linear fades between a set of points via an API which allows one to attach events to properties via a timeline API.  We'll write it with the assumption that we later want to support more complex automation types, for example exponential fades, but avoid doing that for now for the sake of simplicity.  There's a relatively simple path to making linear fades at the block level happen, but sample-accurate fades require significant refactors; put another way, this will be good enough for things like ADSR envelopes or fading out music over time.

Unlike WebAudio, though, we'll have this happen through the user constructing and building a timeline relative to 0, then applying it all at once.  This allows atomic updates and gets around the need to make sure to schedule the beginning of the automation events far enough in the future that your code doesn't "run over" the first one.  While the WebAudio implementation is neat and more flexible, it's also more difficult to use.

Initially, we'll ad this and test this via the Rust bindings.  Adding it to Python is going to be a headache at best: we have no convenient way to just tell every property to get a function and probably someone is going to have to manually hand-write all of them.

This is needed for #17, which is nearly useless without the ability to do envelopes.

CC @neurrone who I think asked for this via another avenue; if not, feel free to unsubscribe.

**Comment by ahicks92** [comment]
The Synthizer side of this is done enough that you can do fadeouts and musical instruments and stuff.  Working C example [here](https://github.com/synthizer/synthizer/blob/2830ccd787c42ad3f3b4b91b2c5c255773fff550/examples/simple_automation.c).  This isn't done primarily because of documentation, but I might try to fit a bit more in here because automating d3 and d6 properties would be stupidly useful.

**Comment by Keithcat1** [comment]
For Python, you could maybe make properties objects so you can call methods on them and define each property type as its own class, and each property type would have a getter / setter property called value that lets you set its value. Instead of this:
generator.gain = 1.5
You'd do:
generator.gain.value = 1.5
I know, API breakage. This would however allow you to:
generator.set_automation_timeline(generator.gain, ...)

**Comment by ahicks92** [comment]
yeah, potentially that.  Alternatively some sort of `gain_automater`.  I think there's a few shortcut ways to specifically punt maintaining the Python bindings issue out one more time if we say "okay well this sucks but we're getting a slightly subpar API" or something.

I'm doing Python last though--this will land in Rust first.  And also the C side is probably going to get a lot of features before I move on from this project, because I can think of lots of things to add and now that the initial work here is done that's not so hard to do.  Once the final API is there we'll see what everything looks like.

**Comment by ahicks92** [comment]
Ok, so the final version of this (in progress, I don't know for sure when) will let you:

- Set/clear automation on any number of properties for any number of objects in one go.
  - Specifically for double, d3, and d6.  Yes, you will be able to do block-accurate movement.
- Ask Synthizer to send you events at specific times, so that you can e.g. arrange to know 50ms or so in advance of the automation running out.
  - Though admittedly that's probably more like 100ms.  More precise synchronization than that is sorta out of scope.
- In order to make this work, expose the ability to get object-local time (a concept which will be documented, but think "if it's paused, time stops advancing").
- Let you tell objects to stop using object-local time for automation and to defer to the context's idea of the current time.
  - But for reasons, not the other way--once this is on, it's on.

Combined with some other stuff that's coming at some point before 1.0 this is enough to, e.g., do simple musical synthesis to the level of most mid-range musical keyboards.

**Comment by ahicks92** [comment]
The final API looks like [this](https://github.com/synthizer/synthizer/blob/d3d577ee40648f4e9ff964d5c74115230ba61369/examples/simple_automation.c).  It's very verbose, but also powerful enough to do anything we might ever want out of it.  I anticipate that everyone will build custom utilities around it to do whatever they need.  There's a few command types left unimplemented, but they're coming.

**Comment by chrisnorman7** [comment]
@ahicks92 This is great stuff!

Just so you know, there's a typo in the argument list of `syz_automationClear`, where you've put an extra "e" in object.

**Comment by ahicks92** [comment]
Thanks, I'll fix it.

**Comment by ahicks92** [comment]
@chrisnorman7 
`syz_automationClear` is actually not a thing and hasn't been for a while; you may be working off an old example.

Saw you were trying to bind this to dart.  it probably won't change again but I'd suggest maybe waiting a week or two until it lands in master.

**Comment by chrisnorman7** [comment]
As far as I can tell, I'm building from the latest version, on the master
branch:

```shell
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

Also, the first entry from `git log`:

```shell
commit 59568d300d23dd2b3aa89cee5028c950eb89ee8d (HEAD -> master,
origin/master, origin/HEAD)
Author: Austin Hicks ***@***.***>
Date:   Sun Sep 12 13:44:20 2021 -0700

    Actually deinit events in the example to try to uncover #75
```

Take care,

Chris Norman

**Comment by ahicks92** [comment]
yeah, the automation branch changes the entire API.  You probably in general want to wait and bind actual releases.

**Comment by chrisnorman7** [comment]
Oh woops. Thank you.

I did what you suggested and abstracted my Synthizer stuff away anyways, so
it won't be too much of a headache.

Cheers,

Take care,

Chris Norman

**Comment by ahicks92** [comment]
Ok. The actual real version of this just landed on master.  I reserve the right to change it before the next release, but that's not likely to happen.  See examples/simple_automation.cpp and examples/automation_circle.cpp.  in addition to the obvious thing you can apply it to d3 and d6 properties which is what the latter shows.  The two remaining things are documentation (easy, probably in the next week) and some accuracy issues around the first couple blocks of audio (significantly harder, we'll do a release without it because this covers the common cases, just not the "I made a keyboard" ones where sharp attacks are necessary).  In other words it's good enough to do things like music fades, and accurate sound animation or whatever.

Also I guess `syz_automationBecomeContextRelative` is missing, I'll add that tomorrow or something.

**Comment by chrisnorman7** [comment]
Great stuff, thank you. I'll get to work.

Take care,

Chris Norman

**Comment by chrisnorman7** [comment]
Also, although `SYZ_OBJECT_TYPE.SYZ_OTYPE_AUTOMATION_TIMELINE` is still present, there is no `SYZ_AutomationTimeline` anymore. Is this a bug?

**Comment by chrisnorman7** [comment]
Sorry for all the questions / problems. 🥇 

Is it necessary to have a time member of `syz_AutomationCommand`, and `syz_AutomationPoint`? Both seem to be used in the simple example, but I'm unclear which value does what, and I'm not sure if it's a case of repeated values.

**Comment by chrisnorman7** [comment]
Ah, ignore me. I've just figured out what the values in `syz_AutomationPoint` actually do.

**Comment by ahicks92** [comment]
yeah, sounds like someone forgot to drop the old object type.  I'll make sure that's not screwy.

The 2 second version is: the command specifies the kind of command, the time on the command specifies when, the variants of the union specify what.  it's always source, time, type, and then varying parameters based on type (for now, either the parameter for the event, or the new value of the property and how to move from the old one).  And by default time is relative to the object, e.g. when you make a generator it starts at 0 and stays at 0 until a source is driving it forward.  The manual is this, but going on for probably 1000 words or something.

**Comment by ahicks92** [comment]
Would it help if I renamed `syz_AutomationPoint` to something else, do you think?  Maybe that'd be worth it.

**Comment by chrisnorman7** [comment]
MMM, not sure. Lemme think about that.

I suspect not. I'm sure in the old version, the first value to
automation point's values was the time at which something would happen,
then the second was the value, and it only occurred to me why there's 6
values when I started reading the circle automation example, ...
AutomationValue though, maybe?

Are you planning to use automation points anywhere else? If not, you could
always chuck `values` and `interpolation_type` onto `
syz_AutomationAppendPropertyCommand`.

BTW, there *might* be a bug with multiple fades. This is only speculative
at this point: Some code I had working before now no longer works, and it
seems to be after the second fade, but I need to do some more hunting. I
just thought I'd give you a heads up.

Take care,

Chris Norman

**Comment by chrisnorman7** [comment]
Bug found.

If you change your simple example to fade the sound back up again, on my system at least, no fade is applied. I even tried putting a clear all properties command in there for good measure, but no dice.

**Comment by ahicks92** [comment]
Can you be more specific?  Are you saying that after the timeline finishes, adding more points doesn't work?

**Comment by chrisnorman7** [comment]
Yes, exactly that.

It fades to 0, which is what I told it to do, but then future fades do
nothing. The gain stays at 0.

Take care,

Chris Norman

**Comment by ahicks92** [comment]
That's probably an easy fix.

To provide some form of advance notice we need to better define what happens when you set properties normally with automation going on at the same time.  I don't know for sure what I want to do there yet other than "something".

**Comment by ahicks92** [comment]
I have pushed an entirely untested fix because it's a weeknight and I don't have time to test the untested fix.  So.  Feel free to test, if you don't I'll get to it this weekend.

**Comment by chrisnorman7** [comment]
Yeah, works fine now mate. Thank you.

**Comment by ahicks92** [comment]
@chrisnorman7 
I just landed what should be close to the final form of this on master, where property writes themselves go via the automation interface as well, so that e.g. if you try to combine automation and property sets it'll do something at least reasonable (specifically insert interpolation type none points at the block where the set lands).  I don't suggest using that functionality but I'm bringing it up because I'm hoping to get you to play with it some.  You don't need to do anything, no API changes or anything, just a lot of reworked internals.

There's a couple functions left that aren't essential and at some point I need to improve the accuracy but I think this is the finish line.

**Comment by chrisnorman7** [comment]
Hi mate,
Good work. Having everything use the same interface makes a lot of sense.

I'm afraid though, having done a fade out, I can't then fade in the same
generator. It just goes straight to the end of the fade. Also, I think
there *might* be something buggy with fade outs, but that could just as
easily be my code (although it worked before). I'm going to attempt to fix
my side, but in the meantime, if you make your simple automation example
fade back in again, on my system at least, that currently does not work.

Take care,

Chris Norman

**Comment by ahicks92** [comment]
I assume you know that fades aren't always relative to zero, and are relative to something like "the current time" (which you will be able to query shortly)?  failing that I'll look this weekend.  Feel free to link examples if they're in your GitHub repo.  it's probably me somehow, but this time isn't a quick 3 line fix: if normal property sets work, timelines are at least partially working.  But for clarification, this:

```
# set a timeline from 0 seconds to 3 seconds
time.sleep(6.0)
# set a timeline from 3 to 6 seconds.
```

Will just jump to the end of the second timeline, since you waited around for 6 seconds before setting a 3 to 6 second timeline.  I don't think you're doing that, but if you are doing that, it's incorrect usage.  In any case you'll be able to get the time you're supposed to be relative to after this weekend, if you're not trying to intentionally build a queue of events like the automation_circle.c example is doing.

**Comment by chrisnorman7** [comment]
Ah, OK, I didn't know they weren't relative to 0. Thank you for that.

Seems they do work, I was just sleeping like you said. That also explains
why my tests are working, but my game is not.

Looking forward to the time-getting method. :)

Take care,

Chris Norman

**Comment by ahicks92** [comment]
As a bridge to me producing docs, there's basically two usage patterns this is for:

- You want to do "animation".  You build up events relative to zero or some time that you saved and keep appending to the queue and bumping the time that you saved forward.  Then Synthizer tracks where it's supposed to be even if your app does the sets early or even a little bit late (though there is a quality drop if it's late).  This is the automation_circle.c example, which uses the event to know when the timeline is empty and the counter to know what time it should push next, even if the events are now in the past.
- You want to do fades and things like that.  You get the current time of the object (pending work) and then schedule relative to that, say 10ms in the future, so that your app has time to complete the calls before the timeline starts being hit.

One way to visualize this is as a graph of the value of the property over time, and a cursor moving from left to right across it that represents the current time: only things after the cursor on the graph are played, but you can still change the graph before the cursor (which will eventually be more useful than it is today).  If you change one point on either side of this cursor then you might change the linear fade that the cursor is doing, but it never jumps back in time, just keeps evaluating the function.  The API is unfortunate in that it has to kind of imply that there's events going on instead of you "drawing" a graph, but eh, what can you do.

**Comment by chrisnorman7** [comment]
That's fine. I'm now keeping track of when my game starts to run, and using
that to get current time. Seems to be working for in-game fades, but I'm
not sure if that'll start to fall away from Synthizer's notion eventually.

Take care,

Chris Norman

**Comment by chrisnorman7** [comment]
Not sure whether to make a new issue for this or not, but it seems I can't set properties on reverbs.

I'm not sure if this is my code or not, but there doesn't appear to be a reverb example I can test with.

**Comment by ahicks92** [comment]
It's almost certainly this issue.  I'll play with it.  I've got next week off so hopefully I can iron out the rest of this stuff really fast.  I wish Synthizer could be unit tested without me spending a man-week on it.

Thanks for testing.  I appreciate it.

**Comment by ahicks92** [comment]
Actually I've been thinking about this and reverb alone shouldn't be broken unless ScalarPannedSource also is.  Which properties are you trying to change?  For a quick and dirty test, try setting t60 to an absurdly large value (say 100) and then back down, see if that does anything.  Though if this is previously working code that you didn't write after these changes it's probably not a test that will tell us much useful.

**Comment by chrisnorman7** [comment]
Hi,
I just wrote a panning example, and it's working fine.

I tried with the reverb, but the value is definitely not getting set. I'm
printing the values, and I'd have thought a value of `100` would be pretty
damn noticeable even if it was somehow printing the wrong values, but no
dice.

That said, clearly something is or should be happening, because if I
set `meanFreePath`
to 50.5, I get a value of range error from Synthizer.

Hopefully that tells you something useful haha.

Take care,

Chris Norman

**Comment by ahicks92** [comment]
Actually I figured out why reverb is broken, and you may now have another undertested quick weeknight fix (it definitely works but this was a 2 minute project, so maybe it also explodes).

For the curious: objects must tick automation timelines now or their properties don't work.  Effects didn't. Now they do.

**Comment by ahicks92** [comment]
I have added `SYZ_P_CURRENT_TIME` to read the current time for any object supporting automation, and `SYZ_P_SUGGESTED_AUTOMATION_TIME` to read where Synthizer suggests that you'd want to start your automation at if it's not going to be late (you always need some offset there, and the hope is that we can make the algorithm more complicated than simple offsets later, thus the property).
